# Assignment #1

# Question #1
a) b) Previously, we obtained the following stationary points of $F$ (the subscripts here do not indicate spatial coordinates): 

$$ x_1^* = (0, 0, 0)$$
$$ x_2^* = ( \sqrt{6} / 4 , \sqrt{6} / 4,  \sqrt{3} / 2)$$
$$ x_3^* = (-\sqrt{6} / 4, -\sqrt{6} / 4, -\sqrt{3} / 2)$$

along with the Hessian Matrix of $F$: 

$$ 
G(x_1, x_2, x_3) = \begin{pmatrix}
    24x_1^2 & -3 & 0  \\ 
    -3      &  6 & -6 \\
    0       & -6 & 12 \\ 
    \end{pmatrix}
$$

To classify these stionary points, we use the well-known [second partial derivative test](https://en.wikipedia.org/wiki/Second_partial_derivative_test). Namely: 

* If $G(x^*)$ is positive definite (equivalently, if all of the eigenvalues are positive), then $F$ attains a local minimum at $x^*$.  
* If $G(x^*)$ is negative definite (equivalently, if all of the eigenvalues are negative), then $F$ attains a local maximum at 
$x^*$. 
* If $G(x^*)$ has both positive and negative eigenvalues, then $x^*$ is a saddle point. 

Our code implementation is as follows: 


In [30]:
import numpy as np 

# Stationary points of F 
stationary_points = [
    [0,0,0], 
    [ np.sqrt(6)/4,  np.sqrt(6)/4,  np.sqrt(3)/2 ], 
    [-np.sqrt(6)/4, -np.sqrt(6)/4, -np.sqrt(3)/2 ]
]

def hessian(x : list): 
    """ Return the Hessian Matrix of F evaluated at x. """
    return np.matrix([
        [24*x[0]**2, -3, 0], 
        [-3, 6, -6], 
        [0, -6, 12]])

for stationary_point in stationary_points: 
    
    print("--------------")
    print("Stationary Point:")
    print(stationary_point) 

    # Obtain the eigenvalues of the Hessian evaluated at that stationary point 
    eigenvals, _ = np.linalg.eig(hessian(stationary_point))

    print("Eigenvalues of the Hessian at Critical Point:")
    print(eigenvals)

    # Perform second partial derivative test 
    if(all(eigenval > 0 for eigenval in eigenvals)):
        print("F attains a minimum!")
    elif(all(eigenval < 0 for eigenval in eigenvals)):
        print("F attains a maximum!")
    else: 
        print("F attains a saddle point!")
    
    print("--------------")

--------------
Stationary Point:
[0, 0, 0]
Eigenvalues of the Hessian at Critical Point:
[-1.75267633  3.88284108 15.86983525]
F attains a saddle point!
--------------
--------------
Stationary Point:
[0.6123724356957945, 0.6123724356957945, 0.8660254037844386]
Eigenvalues of the Hessian at Critical Point:
[ 1.41324613  9.50234757 16.0844063 ]
F attains a minimum!
--------------
--------------
Stationary Point:
[-0.6123724356957945, -0.6123724356957945, -0.8660254037844386]
Eigenvalues of the Hessian at Critical Point:
[ 1.41324613  9.50234757 16.0844063 ]
F attains a minimum!
--------------


We conclude that $x_1^*$ is a saddle point, $x_2^*$ is a minimum, and $x_3^*$ is also a minimum. 

# Question #3
c) Given the search direction: 

$$  p_k := p(\hat{x}_k) = 
\begin{pmatrix}
-2\hat{x}_{1,k} + \hat{x}_{2,k} + 2 \\ 
-2\hat{x}_{2,k} + \hat{x}_{1,k} + \hat{x}_{3,k} \\ 
-2\hat{x}_{3,k} + \hat{x}_{2,k} + 8 
\end{pmatrix}
$$

Our steepest descent algoritm, $\hat{x}_{k+1} = \hat{x}_k + \alpha p_k$  , is implemented below. 

In [31]:
import numpy as np 

def objective_func(x : np.ndarray) -> float:
    """ The objective function to be minimized """
    return x[0]**2 + x[1]**2 + x[2]**2 - x[0]*x[1] - x[1]*x[2] - 2*x[0] - 8*x[2]

def search_direction(x_hat : np.ndarray) -> np.ndarray: 
    """ Return the search direction for our objective function, for a given input"""
    return np.array([
        -2*x_hat[0] + x_hat[1] + 2, 
        -2*x_hat[1] + x_hat[0] + x_hat[2], 
        -2*x_hat[2] + x_hat[1] + 8  
    ])

def steepest_descent(x0 : np.ndarray, alpha : float, maxiter=1000, tol=10e-6) -> dict:
    """ Perform steepest descent algorithm to find the minimum of our objective function """

    x_curr = np.array([0, 0, 0]) 
    x_prev = np.array([0, 0, 0])

    for iter in range(maxiter): 
        if(iter == 0): 
            x_curr = x0 + alpha*search_direction(x0)
        else: 
            x_curr = x_prev + alpha*search_direction(x_prev)
        
        # Check convergence 
        if(np.linalg.norm(x_curr - x_prev, ord=np.inf) <= tol):
            break 
        
        # Perform update 
        x_prev = x_curr
    
    return {
        "x"    : x_curr, 
        "iter" : iter + 1, 
        "val"  : objective_func(x_curr) 
    }

# Initial guess vector
x0 = np.array([1, 1, 1])

# alphas interested in evaluating 
alphas = [ 0.1, 0.2, 0.5, 1.0 ]

# Perform the steepest descent algorithm for each alpha value
for alpha in alphas: 

    res = steepest_descent(x0, alpha) 

    print("------------------------------------")
    print("Alpha:                  " + str(alpha))
    print("Current Function Value: " + str(res["val"]))
    print("Iterations:             " + str(res["iter"]))
    print("Minimum Computed:       " + str(res["x"]))
    print("------------------------------------")


------------------------------------
Alpha:                  0.1
Current Function Value: -29.49999998522613
Iterations:             171
Minimum Computed:       [3.4998877  4.99984119 6.4998877 ]
------------------------------------
------------------------------------
Alpha:                  0.2
Current Function Value: -29.499999996818843
Iterations:             89
Minimum Computed:       [3.49994789 4.99992631 6.49994789]
------------------------------------
------------------------------------
Alpha:                  0.5
Current Function Value: -29.499999999883585
Iterations:             37
Minimum Computed:       [3.49999237 4.99998474 6.49999237]
------------------------------------
------------------------------------
Alpha:                  1.0
Current Function Value: nan
Iterations:             1000
Minimum Computed:       [nan nan nan]
------------------------------------


/tmp/ipykernel_47212/876630572.py:10: RuntimeWarning: overflow encountered in scalar add
  -2*x_hat[0] + x_hat[1] + 2,
/tmp/ipykernel_47212/876630572.py:11: RuntimeWarning: overflow encountered in scalar multiply
  -2*x_hat[1] + x_hat[0] + x_hat[2],
/tmp/ipykernel_47212/876630572.py:12: RuntimeWarning: overflow encountered in scalar add
  -2*x_hat[2] + x_hat[1] + 8
/tmp/ipykernel_47212/876630572.py:25: RuntimeWarning: invalid value encountered in add
  x_curr = x_prev + alpha*search_direction(x_prev)


Indeed, we find that for $\alpha = 0.1$ our algorithm converges in 171 iterations, for $\alpha = 0.2$ our algorithm converges in 89 iterations, and for $\alpha = 0.5$ our algorithm converges in 37 iterations. Unfortunatlely, our algorithm fails to converge for $\alpha = 1$, as we exceed the maximum number of iterations and do not reach the minimum. Thus, $\alpha = 0.5$ gives the best performance overall, which agrees with our answer for part b of this problem. 

d) Here, rather than using MATLAB's `fminunc`, we use `scipy.optimize.minimize` with  `method="BFGS`. 

In [32]:
from scipy.optimize import minimize

# Minimize using BFGS method 
res = minimize(objective_func, x0, method='BFGS', options={'gtol': 1e-6, 'disp' : True})

print("Minimum Computed:")
print(res.x)


Optimization terminated successfully.
         Current function value: -29.500000
         Iterations: 9
         Function evaluations: 40
         Gradient evaluations: 10
Minimum Computed:
[3.5000001  5.00000025 6.49999985]


Unsurprisingly, `scipy.optimize.minimize` outperforms our own steepest descent algorithm, as it converges to the same minimum but in 9 iterations rather than 37, which implies a performance gain of a factor of roughly four. 

# Question #4
c) In part b), we invoked the fact that: 

$$ \Delta F \approx -\lambda^{\text{T}} \Delta f = -\lambda_1 \Delta f_1 - \lambda_2 \Delta f_2 - \lambda_3 \Delta f_3 $$

to argue that for a perturbation in the constraint $f_1$, we expect the difference in the optimal cost to be: 

$$ \Delta F \approx - \lambda_1 \Delta f_1 = -\lambda_1 (f_1^\prime - f_1) = -\lambda_1 (0.1) $$

To confirm this result, let us first solve the constrained minimization problem at hand, using `scipy.optimize.minimize` with `method="SLSQP"` (see `examples\unconstrained_minimization.ipynb` for a more detailed example of how to do this). Our implementation is as follows: 

In [33]:
import numpy as np 
from   scipy.optimize import minimize

def func(x : list) -> float:
  """ The objective function to be minimized """
  return x[0]**2 + x[1]**2 - 6*x[0]*x[1] - 4*x[0] - 5*x[1] 

# Constraint functions 
def con_1(x : list) -> float:
  """ First constraint: -2x_1 + x_2 + 1 >= 0 """
  return -2*x[0] + x[1] + 1  

def con_2(x : list) -> float: 
  """ Second constraint: -x_1 - x_2 + 4 >= 0 """
  return -x[0] - x[1] + 4

def con_3(x : list) -> float: 
  """ Third constraint: x_1 + 1 >= 0 """
  return x[0] + 1 

# Define the constraint object 
ineq_cons = {'type' : 'ineq', 'fun' : lambda x : np.array([con_1(x), con_2(x), con_3(x)])} 

# Initial guess vector
x0  = [0, 0]

# Minimize using SLSQP method 
res = minimize(func, x0, method='SLSQP', 
                         constraints=ineq_cons, 
                         options={'disp' : True})
print("Minimum Computed:")
print(res.x)

Optimization terminated successfully    (Exit mode 0)
            Current function value: -33.44444444635195
            Iterations: 3
            Function evaluations: 9
            Gradient evaluations: 3
Minimum Computed:
[1.66666667 2.33333333]


Thus the solution to our constrained minimization problem, is $x^* = [5/3, 7/3]^{\text{T}}$ and the values of the constraints are: 




In [34]:
print("Constraint Values")
print(con_1(res.x))
print(con_2(res.x))
print(con_3(res.x)) 

Constraint Values
-7.446443461844865e-11
-1.5282619614254145e-10
2.6666666667424304


Since the values of the first and second constraint are (essentially) 0, we conclude that these are active constraints and the third constraint is innactive. Thus, $\lambda_1 \geq 0, \lambda _2 \geq 0, \lambda_3 = 0$. Substituting these results into our equations for $\frac{\partial L}{\partial x}$ and 
$\frac{\partial L}{\partial y}$ gives: 

$$ 2\lambda_1 + \lambda_2 = 4 - 2(5/3) + 6(7/3) = 44/3$$
$$ -\lambda_1 + \lambda_2 = 5 - 2(7/3) + 6(5/3) = 31/3$$ 

So, $\lambda_1 = 13/9, \lambda_2 = 106/9$ and we estimate: 

$$ \Delta F \approx - (13/9) \times (0.1) \approx -0.14440 $$

Our verification of this is as follows: 

In [35]:
def con_1_mod(x : list) -> float:
  """ First constraint modified: -2x_1 + x_2 + 1.1 >= 0 """
  return -2*x[0] + x[1] + 1.1

# Define the modified constraint object 
ineq_cons_mod = {'type' : 'ineq', 'fun' : lambda x : np.array([con_1_mod(x), con_2(x), con_3(x)])} 

# Initial guess vector
x0  = [0, 0]

# Minimize using SLSQP method 
res_mod = minimize(func, x0, method='SLSQP', constraints=ineq_cons_mod, options={'disp' : True})

print("--------------------------")
print("Cost Function Difference: ")
print(str(res_mod.fun - res.fun))

Optimization terminated successfully    (Exit mode 0)
            Current function value: -33.580000000345905
            Iterations: 3
            Function evaluations: 9
            Gradient evaluations: 3
--------------------------
Cost Function Difference: 
-0.13555555399395303
